### PR TITLE
T2D-185 Review tests

### PR DESCRIPTION
**Pull request by jorizci**
Depends on #36 
* Added Cucumber as dependency
* Added submodule with common test functionality
* Added integration test for jpa and mongo

### CODE REVIEW

File: accession-commons-jpa/src/test/resources/features/modification.feature [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  Scenario Outline: I want to update the fields that identify uniquely an object.
+  without losing access to the old information.
+    Given already accessioned <objects>
+    When user updates <accession> patch <patch> with <input>
+    Then user should receive version list for <accession> with 1 version
+    And hash of version 1 should be <hash>
+    Examples:
+      | objects | input | accession    | patch | hash                                     |
+      | A       | AA    | id-service-A | 1     | 801C34269F74ED383FC97DE33604B8A905ADB635 |
```
Comment (selvaebi) [inline accession-commons-jpa/src/test/resources/features/modification.feature]: can we have Given hash as well , to show the hash has changed ?

File: accession-commons-jpa/src/test/resources/features/modification.feature [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      | A,B,C,E | AA    | id-service-A |
+
+  Scenario: I want to create a new patch of a non existing object
+    When user sends patch AA for accession id-service-A
+    Then user receives accession does not exist
+
+  Scenario Outline: I want to update the fields that identify uniquely an object.
+  without losing access to the old information.
+    Given already accessioned <objects>
+    When user updates <accession> patch <patch> with <input>
```
Comment (selvaebi) [inline accession-commons-jpa/src/test/resources/features/modification.feature]: can we have <accession> version <version> instead of "patch" please, to match with "Then" output

File: accession-commons-jpa/src/test/resources/features/modification.feature (hunk 1 above)
Comment (jorizci) [reply accession-commons-jpa/src/test/resources/features/modification.feature]: You mean the hash of the original object? Maybe I can stretch the language a bit, but the idea is that we are on the test  case of a set of updates and the information is already loaded so the user accion would be do the update not checking that the current element has hash X

File: accession-commons-jpa/src/test/resources/features/modification.feature (hunk 2 above)
Comment (jorizci) [reply accession-commons-jpa/src/test/resources/features/modification.feature]: I am more keen to use patch in both that version.

File: accession-commons-jpa/src/test/resources/features/modification.feature (hunk 1 above)
Comment (selvaebi) [reply accession-commons-jpa/src/test/resources/features/modification.feature]: Ya i understand, i  thought the then was showing that the hash has changed , but the main use case is version not changed.

File: accession-commons-jpa/src/test/resources/features/modification.feature [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      | objects | input | accession    | patch | hash                                     |
+      | A       | AA    | id-service-A | 1     | 801C34269F74ED383FC97DE33604B8A905ADB635 |
+      | B       | BB    | id-service-B | 1     | 71C9DB717578B9EE49A59E69375C16C0627DFFEF |
+      | A,B,C,E | AA    | id-service-A | 1     | 801C34269F74ED383FC97DE33604B8A905ADB635 |
+
+  Scenario: I want to update the fields that identify uniquely an object, but we already have an object with the
+  same values.
+    Given already accessioned A,B
+    When user updates id-service-A patch 1 with B
+    Then user receives an unknown error
```
Comment (selvaebi) [inline accession-commons-jpa/src/test/resources/features/modification.feature]: the Then may be "user receives hash/object clash during update error " instead of unknown error ?

Comment (jorizci) [reply accession-commons-jpa/src/test/resources/features/modification.feature]: Here comes the kicker, the exception is called Unknown error.

**Comment by selvaebi** [comment]
Don't we need a feature for deprecation ?

File: accession-commons-jpa/src/test/resources/features/modification.feature (hunk 3 above)
Comment (jorizci) [reply accession-commons-jpa/src/test/resources/features/modification.feature]: Yup

**Comment by jorizci** [comment]
We do need a feature for deprecation and history, they were not done as the changes in history affected them.

File: accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/commons/accession/AccessioningServiceStepDefinitions.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    public void patchMustHaveVersionsIncreased() {
+        tester.getLastMultipleVersionResult().assertVersionAreIncreased();
+    }
+
+    @Then("^user receives accession does not exist$")
+    public void userReceivesAccessionDoesNotExistError() {
+        tester.getLastMethodResponse().assertThrow(AccessionDoesNotExistException.class);
+    }
+
+    @When("^user updates ([\\w-]+) patch (\\d+) with ([\\w-]+)$")
```
Comment (cyenyxe) [inline accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/commons/accession/AccessioningServiceStepDefinitions.java]: We should homogeneize the vocabulary, either using version or patch but not both (see version in previous methods).

File: accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/commons/accession/AccessioningServiceStepDefinitions.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        tester.update(accession, patch, newData);
+    }
+
+
+    @And("^hash of version (\\d+) should be ([\\w-]+)$")
+    public void hashOfVersionShouldBeHash(int version, String hash) {
+        tester.getLastMultipleVersionResult().assertHash(version, hash);
+    }
+
+    @When("^user merges ([\\w-]+) with ([\\w-]+) reason: ([\\w ]+)$")
```
Comment (cyenyxe) [inline accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/commons/accession/AccessioningServiceStepDefinitions.java]: `user merges ([\\w-]+) into ([\\w-]+) due to: ([\\w ]+)`

File: accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/commons/accession/AccessioningServiceStepDefinitions.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    @When("^user retrieves objects: ([\\w-,]+)$")
+    public void userRetrievesObjectA(String values) {
+        tester.get(toModels(values));
+    }
+
+    private List<TestModel> toModels(String values) {
+        return Arrays.stream(values.split(",")).map(TestModel::of).collect(Collectors.toList());
+    }
```
Comment (cyenyxe) [inline accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/commons/accession/AccessioningServiceStepDefinitions.java]: This method should be moved right after the `@Given` at the beginning of the class.

File: accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/test/configuration/TestJpaDatabaseServiceTestConfiguration.java [hunk 7]
```diff
@@ -50,6 +53,9 @@
 @EnableJpaRepositories(basePackages = {"uk.ac.ebi.ampt2d.test.persistence",
         "uk.ac.ebi.ampt2d.commons.accession.persistence.jpa.repositories"
 })
+@AutoConfigureCache
+@AutoConfigureDataJpa
+@AutoConfigureTestEntityManager
```
Comment (cyenyxe) [inline accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/test/configuration/TestJpaDatabaseServiceTestConfiguration.java]: Why not [DataJpaTest](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/test/autoconfigure/orm/jpa/DataJpaTest.html) instead of these 3 annotations?

File: accession-commons-jpa/src/test/resources/features/modification.feature (hunk 2 above)
Comment (cyenyxe) [reply accession-commons-jpa/src/test/resources/features/modification.feature]: Agree with favoring "patch", it is the term we have agreed to use. "version" is just the way the field is serialized into the database because there is no way around it.

File: accession-commons-jpa/src/test/resources/features/modification.feature [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      | A       | AA    | id-service-A |
+      | B       | BB    | id-service-B |
+      | A,B,C,E | AA    | id-service-A |
+
+  Scenario: I want to create a new patch of a non existing object
+    When user sends patch AA for accession id-service-A
+    Then user receives accession does not exist
+
+  Scenario Outline: I want to update the fields that identify uniquely an object.
+  without losing access to the old information.
```
Comment (cyenyxe) [inline accession-commons-jpa/src/test/resources/features/modification.feature]: `object, without` (note the comma instead of dot)

File: accession-commons-jpa/src/test/resources/features/modification.feature [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    When user updates <accession> patch <patch> with <input>
+    Then user should receive version list for <accession> with 1 version
+    And hash of version 1 should be <hash>
+    Examples:
+      | objects | input | accession    | patch | hash                                     |
+      | A       | AA    | id-service-A | 1     | 801C34269F74ED383FC97DE33604B8A905ADB635 |
+      | B       | BB    | id-service-B | 1     | 71C9DB717578B9EE49A59E69375C16C0627DFFEF |
+      | A,B,C,E | AA    | id-service-A | 1     | 801C34269F74ED383FC97DE33604B8A905ADB635 |
+
+  Scenario: I want to update the fields that identify uniquely an object, but we already have an object with the
```
Comment (cyenyxe) [inline accession-commons-jpa/src/test/resources/features/modification.feature]: `but another object already has the same values`

File: accession-commons-mongodb/src/test/java/uk/ac/ebi/ampt2d/commons/accession/AccessioningServiceStepDefinitions.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+    //Required for nosql unit
+    @Autowired
+    private ApplicationContext applicationContext;
+
+    @Autowired
+    private AccessioningServiceTester tester;
+
+    @Before
+    public void doSomethingBefore() {
```
Comment (cyenyxe) [inline accession-commons-mongodb/src/test/java/uk/ac/ebi/ampt2d/commons/accession/AccessioningServiceStepDefinitions.java]: The convention name is `setUp` :stuck_out_tongue:

File: accession-commons-mongodb/src/test/java/uk/ac/ebi/ampt2d/test/configuration/MongoDbTestConfiguration.java [hunk 11]
```diff
@@ -80,7 +80,8 @@ public TestMongoDbInactiveAccessionService testMongoDbInactiveAccessionService()
     @Bean
     public AccessioningService<TestModel, String, String> testMongoDbAccessioningService() {
         return new BasicAccessioningService<>(
-                new SingleAccessionGenerator<>(o -> "id-" + o.getValue()),
+                new SingleAccessionGenerator<>(o ->
+                        "id-" + o.getValue()),
```
Comment (cyenyxe) [inline accession-commons-mongodb/src/test/java/uk/ac/ebi/ampt2d/test/configuration/MongoDbTestConfiguration.java]: Maybe you didn't intend this line to be split, at least I find it harder to read.

File: accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/AccessionVersionWrapperTester.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+            ThrowingSupplier<AccessionVersionsWrapper<TestModel, String, String>> functionCall) {
+        super(functionCall);
+    }
+
+    public AccessionVersionWrapperTester assertTotalVersions(int total) {
+        assertEquals(total, getData().getModelWrappers().size());
+        return this;
+    }
+
+    public AccessionVersionWrapperTester assertVersionAreIncreased() {
```
Comment (cyenyxe) [inline accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/AccessionVersionWrapperTester.java]: "versions are" or "version is"

File: accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/HistoryTester.java [hunk 13]
```diff
@@ -66,23 +68,23 @@ public HistoryAccessionTester assertEvent(int i, Consumer<HistoryEvent<TestModel
     }
 
     public static Consumer<HistoryEvent<TestModel, String>> assertAccession(String accession) {
-        return event -> assertEquals(accession, event.getAccession());
+        return event -> Assert.assertEquals(accession, event.getAccession());
```
Comment (cyenyxe) [inline accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/HistoryTester.java]: Is this change really needed, given that the static import points to the method directly?

File: accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/AccessioningServiceTester.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+import uk.ac.ebi.ampt2d.commons.accession.core.AccessioningService;
+import uk.ac.ebi.ampt2d.test.models.TestModel;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class AccessioningServiceTester {
```
Comment (cyenyxe) [inline accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/AccessioningServiceTester.java]: Is one of this created every time a scenario is tested, in order to avoid dependencies across tests?

File: accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/test/configuration/TestJpaDatabaseServiceTestConfiguration.java (hunk 7 above)
Comment (jorizci) [reply accession-commons-jpa/src/test/java/uk/ac/ebi/ampt2d/test/configuration/TestJpaDatabaseServiceTestConfiguration.java]: Because DataJpaTest introduces all changes inside a transaction that doesn't play well with the system, as it doesn't expect to be inside another transaction and the behavior breaks completely.

File: accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/HistoryTester.java (hunk 13 above)
Comment (jorizci) [reply accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/HistoryTester.java]: Nope, I was probably fighting with some import.

File: accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/AccessioningServiceTester.java (hunk 14 above)
Comment (jorizci) [reply accession-commons-test/src/main/java/uk/ac/ebi/ampt2d/test/testers/AccessioningServiceTester.java]: Yes, it is a different world for each scenario execution. There are no dependences, and databases are cleaned before each test.
